lib/core/refresh.js

```diff
@@ -0,0 +1,246 @@
+import jwtDecode, { InvalidTokenError } from 'jwt-decode'
+import { getProp } from './utilities'
+import ExpiredAuthSessionError from './includes/ExpiredAuthSessionError'
+
+export class RefreshController {
+  constructor (scheme) {
+    this.scheme = scheme
+    this.refreshPromise = null
+    this.refreshInterval = undefined
+  }
+
+  // Returns a promise which is resolved when refresh is completed
+  // Call this function when you intercept a request with an expired token.
+  // Multiple requests will be queued until the first has completed token refresh.
+  handleRefresh () {
+    // Another request has started refreshing the token, wait for it to complete
+    if (this.refreshPromise) {
+      return this.refreshPromise
+    }
+
+    return this._doRefresh()
+  }
+
+  _doRefresh () {
+    this.refreshPromise = new Promise(resolve => {
+      this.scheme._refreshToken().then(response => {
+        this.refreshPromise = null
+        resolve(response)
+      })
+    })
+
+    return this.refreshPromise
+  }
+
+  initializeAutomaticRefresh () {
+    // If auto refresh is enabled, schedule token refresh
+    if (this.scheme.options.autoRefresh.enable) {
+      this.initializeScheduledRefresh()
+    }
+
+    this.initializeRequestInterceptor()
+  }
+
+  // ---------------------------------------------------------------
+  // Refreshes token on set intervals
+  // Token is refreshed at 75% of the token expiration
+  // Call this function once from your mounted hook, client side
+  // ---------------------------------------------------------------
+  initializeScheduledRefresh () {
+    let intervalDuration = (this.scheme._getTokenExpiration() - Date.now()) * 0.75
+    if (intervalDuration < 1000) {
+      // in case you misconfigured refreshing this will save your auth-server from a self-induced DDoS-Attack
+      intervalDuration = 1000
+    }
+
+    this.refreshInterval = setInterval(() => {
+      this.handleRefresh()
+    }, intervalDuration)
+  }
+
+  // ---------------------------------------------------------------
+  // Watch requests for token expiration
+  // Refresh tokens if token has expired
+  // ---------------------------------------------------------------
+  initializeRequestInterceptor (refreshEndpoint) {
+    this.scheme.$auth.ctx.app.$axios.onRequest(async config => {
+      // Don't intercept refresh token requests
+      if (config.url === refreshEndpoint) {
+        return config
+      }
+
+      // Sync tokens
+      const token = this.scheme.$auth.syncToken(this.scheme.name) || false
```

Comment:
TIP: In class constructor, we can set `this.$auth=this.scheme.$auth` to make code simpler and save few bytes :)

Reply:
Ok!

Reply:
I'll do the same thing for options `this.options=this.scheme.options`

Reply:
For options, I think doesn't make sense (unless getting class specific options as constructor argument) because `this.options` implies class-specific options but this way it refers to entire options.

Reply:
Ok

Reply:
Fixed in commit https://github.com/nuxt-community/auth-module/pull/548/commits/13b419c781dec9d6fb340c7be492283d05c56811